lib/names/workloadendpoint_test.go

```diff
@@ -0,0 +1,200 @@
+			Node:         "node-1",
+			Orchestrator: "k8s",
+			Pod:          "pod-name",
+			Endpoint:     "eth0",
+		}, "node--1-k8s-pod--name-eth0", ""),
+		Entry("Missing  workload but other unknown orchestrator fields are present", names.WorkloadEndpointIdentifiers{
+			Node:         "node-1",
+			Orchestrator: "foo",
+			Pod:          "pod-name",
+			Endpoint:     "eth0",
+		}, "node--1-foo-", ""),
+		Entry("Missing k8s Pod", names.WorkloadEndpointIdentifiers{
+			Node:         "node-1",
+			Orchestrator: "k8s",
+			Endpoint:     "eth0",
+		}, "node--1-k8s-", ""),
+		Entry("CNI container ID starts with a -", names.WorkloadEndpointIdentifiers{
+			Node:         "node-1",
+			Orchestrator: "cni",
+			ContainerID:  "-abcdefa0123456",
+			Endpoint:     "eth0",
+		}, "", "containerID"),
+		Entry("Unknown orchestrator, workload ends with a -", names.WorkloadEndpointIdentifiers{
+			Node:         "node-1",
+			Orchestrator: "foo",
+			Workload:     "foo-foo-",
+			Endpoint:     "abcdefgh",
+		}, "", "workload"),
+		Entry("Node is missing", names.WorkloadEndpointIdentifiers{
+			Orchestrator: "k8s",
+			Pod:          "pod-name",
+			Endpoint:     "eth0",
+		}, "", "node"),
+		Entry("Orchestrator is missing", names.WorkloadEndpointIdentifiers{
+			Node:     "node-1",
+			Pod:      "pod-name",
+			Endpoint: "eth0",
+		}, "", "orchestrator"),
+	)
+
+	DescribeTable("WorkloadEndpoint name matching",
+		func(ids names.WorkloadEndpointIdentifiers, name string, expectValid bool, erroredField string) {
+			valid, err := ids.NameMatches(name)
+			if len(erroredField) != 0 {
+				Expect(valid).To(BeFalse())
+				Expect(err).To(HaveOccurred())
+				Expect(err.(cerrors.ErrorValidation).ErroredFields).To(HaveLen(1))
+				Expect(err.(cerrors.ErrorValidation).ErroredFields[0].Name).To(Equal(erroredField))
+			} else {
+				Expect(valid).To(Equal(expectValid))
+			}
+		},
+		Entry("Node is missing", names.WorkloadEndpointIdentifiers{
+			Orchestrator: "k8s",
+			Pod:          "pod-name",
+			Endpoint:     "eth0",
+		}, "", false, "node"),
+		Entry("Orchestrator is missing", names.WorkloadEndpointIdentifiers{
+			Node:     "node-1",
+			Pod:      "pod-name",
+			Endpoint: "eth0",
+		}, "", false, "orchestrator"),
+		Entry("Valid k8s endpoint and matching name", names.WorkloadEndpointIdentifiers{
+			Node:         "node-1",
+			Orchestrator: "k8s",
+			Pod:          "pod-name",
+			Endpoint:     "eth0",
+		}, "node--1-k8s-pod--name-eth0", true, ""),
+		Entry("Fully qualified k8s endpoint, non-matching Endpoint", names.WorkloadEndpointIdentifiers{
+			Node:         "node-1",
+			Orchestrator: "k8s",
+			Pod:          "pod-name",
+			Endpoint:     "eth0",
+		}, "node--1-k8s-pod--name-eth", false, ""),
+		Entry("CNI endpoint missing container ID and matching name", names.WorkloadEndpointIdentifiers{
+			Node:         "node-1",
+			Orchestrator: "cni",
+			Endpoint:     "eth0",
+		}, "node--1-cni-orchestratorid-eth0", true, ""),
```

Comment:
not sure I understand where this `orchestratorid` string comes from? How is that being placed in place of `ContainerID`?

Reply:
the key thing is that the IDs don't include the ContainerID - so it can be regarded as a wildcard match.  So the name does match that.

Reply:
workflow is:

-  create a WEPIdentifiers with the subset of ids
-  construct the name prefix
-  do a prefix list
-  parse the items in the list and check the name matches the IDs

this final step is necessary because it's possible to get a prefix match that doesn't actually match the required identifiers.  For example (and this would be a real scenario for CNI), suppose you are querying endpoints with node=node1, orch=k8s, pod=pod and endpoints is wild carded.

The name prefix would be   `node1-k8s-pod-`

but that could also return endpoints with a pod call "pod-1", because the name of that endpoints would be `node1-k8s-pod--1-eth0` which matches the prefix.

Reply:
i've beefed up the comment for the WorkloadEndpointIdentifiers doc.